packages/relay/src/lib/services/debugService/index.ts

```diff
@@ -0,0 +1,219 @@
+   */
+  async resolveAddress(
+    address: string,
+    types = [constants.TYPE_CONTRACT, constants.TYPE_TOKEN, constants.TYPE_ACCOUNT],
+    requestIdPrefix?: string,
+  ): Promise<string> {
+    const entity = await this.mirrorNodeClient.resolveEntityType(
+      address,
+      types,
+      'debug_traceTransaction',
+      requestIdPrefix,
+    );
+
+    if (
+      entity &&
+      (entity.type === constants.TYPE_CONTRACT || entity.type === constants.TYPE_ACCOUNT) &&
+      entity.entity?.evm_address
+    ) {
+      return entity.entity.evm_address;
+    }
+
+    return address;
+  }
+
+  /**
+   * Returns the final formatted response for callTracer config.
+   *
+   * @async
+   * @param {string} transactionHash - The hash of the transaction to be debugged.
+   * @param {any} tracerConfig - The tracer config to be used.
+   * @param {string} requestIdPrefix - The request prefix id.
+   * @returns {Promise<object>} The formatted response.
+   */
+  async callTracer(transactionHash: string, tracerConfig: any, requestIdPrefix?: string): Promise<object> {
+    try {
+      const [actionsResponse, transactionsResponse] = await Promise.all([
+        this.mirrorNodeClient.getContractsResultsActions(transactionHash, requestIdPrefix),
+        this.mirrorNodeClient.getContractResultWithRetry(transactionHash),
+      ]);
+
+      const { call_type: type } = actionsResponse.actions[0];
+      const formattedActions = await this.formatActionsResult(actionsResponse.actions, requestIdPrefix);
+
+      const {
+        from,
+        to,
+        amount,
+        gas_limit: gas,
+        gas_used: gasUsed,
+        function_parameters: input,
+        call_result: output,
+        error_message: error,
+        result,
+      } = transactionsResponse;
+
+      const [fromToEvmAddress, toToEvmAddress] = await Promise.all([
+        this.resolveAddress(
+          from,
+          [constants.TYPE_CONTRACT, constants.TYPE_TOKEN, constants.TYPE_ACCOUNT],
+          requestIdPrefix,
+        ),
+        this.resolveAddress(
+          to,
+          [constants.TYPE_CONTRACT, constants.TYPE_TOKEN, constants.TYPE_ACCOUNT],
+          requestIdPrefix,
+        ),
+      ]);
+
+      const value = amount === 0 ? '0x0' : numberTo0x(amount);
+      const errorResult = result !== 'SUCCESS' ? result : undefined;
+
+      return {
+        type,
+        from: fromToEvmAddress,
+        to: toToEvmAddress,
+        value,
+        gas: numberTo0x(gas),
+        gasUsed: numberTo0x(gasUsed),
+        input,
+        output: result !== 'SUCCESS' ? error : output,
+        ...(result !== 'SUCCESS' && { error: errorResult }),
+        ...(result !== 'SUCCESS' && { revertReason: decodeErrorMessage(error) }),
+        // if we have two actions or more executed during the transaction the first one is returned in the top and the other ones in
+        // a calls array, although both are returned form the actions endpoint
```

Comment:
Q: What do you mean by `although both are returned form the actions endpoint`?

Reply:
I have made it more clear now, but what i mean is that in the actions endpoint we return all calls executed during the transaction, however in our response in the sub-calls array we need to return all but the first one